### DISCUSSION TITLE
Unable to visualize `sensor_msgs/msg/Image` stored in MCAP rosbag

### DISCUSSION THREAD

**Discussion by aditya2592** [foxglove/community#951] [Q&A] [answered]
**Description**

I have an MCAP file where an RGB image of type `sensor_msgs/msg/Image` is stored in a topic. When I try to visualize this using the `Image` layout after importing the MCAP file, I get the following error:
```
RangeError: Invalid typed array length: 2682033074
    at new Uint8Array (<anonymous>)
    at CdrReader.uint8Array (file:///snap/foxglove-studio/27/resources/app.asar/renderer/2098.js:1:2766)
    at file:///snap/foxglove-studio/27/resources/app.asar/renderer/2098.js:1:17718
    at MessageReader.readComplexType (file:///snap/foxglove-studio/27/resources/app.asar/renderer/2098.js:1:16780)
    at MessageReader.readMessage (file:///snap/foxglove-studio/27/resources/app.asar/renderer/2098.js:1:15956)
    at Object.deserializer (file:///snap/foxglove-studio/27/resources/app.asar/renderer/1023.js:1:7532)
    at processRecord (file:///snap/foxglove-studio/27/resources/app.asar/renderer/4766.js:1:6970)
    at McapStreamingIterableSource.initialize (file:///snap/foxglove-studio/27/resources/app.asar/renderer/4766.js:1:7255)
    at async McapIterableSource.initialize (file:///snap/foxglove-studio/27/resources/app.asar/renderer/4766.js:1:11355)
    at async WorkerIterableSourceWorker.initialize (file:///snap/foxglove-studio/27/resources/app.asar/renderer/4126.js:1:19821)
```
The same topic when being visualized live through `websocket` works fine. I am able to see the image in Foxglove Studio.

- Version: 1.35.2
- Platform: Ubuntu 22.04

**Steps To Reproduce**

MCAP file was stored with below writer config:

```
noChunkCRC: false
noChunking: false
noMessageIndex: false
noSummary: false
chunkSize: 786432
compression: "Zstd"
compressionLevel: "Fast"
forceCompression: false
```

**Expected Behavior**
Visualize topic of type `sensor_msgs/msg/Image` both live and through the MCAP file

**Comment by snosenzo** [comment]
Thanks for filing this -- would you be able to share an example MCAP file that we can test against?

**Comment by aditya2592** [comment]
[Here](https://drive.google.com/file/d/195A-jeaFjjP3igY6GjOIziQW7U-HxTz_/view?usp=sharing) is a link to a test bag containing only 1 topic with above error

**Comment by jtbandes** [comment]
How did you record this bag? Something has gone wrong in the recording process.

There seems to be a mismatch between the schema name, schema data, and message data. The schema data embedded in the file is the schema for `sensor_msgs/CompressedImage`:

<img width="698" alt="Screenshot 2023-01-05 at 10 53 46 AM" src="https://user-images.githubusercontent.com/14237/210858474-335c333e-d7dd-4ebf-9a99-951f495ff96a.png">

However, the schema name and the actual message data appear to be a `sensor_msgs/Image` — note the presence of the `width` and `height` (640x480) and the `encoding` (`8UC3`) in the message data:
<img width="247" alt="image" src="https://user-images.githubusercontent.com/14237/210858681-83dadc0e-31b6-486c-80e1-8104d6c82c0b.png">

<img width="580" alt="image" src="https://user-images.githubusercontent.com/14237/210858159-33c2cfd7-19d9-450d-9987-00ad89535d0b.png">

**Comment by jtbandes** [comment]
Converting to a discussion because this is not a bug with studio.

**Comment by james-rms** [comment, marked as answer]
@aditya2592 are you able to link the version of `rosbag2_storage_mcap` you used to record this bag?

**Reply by aditya2592** [reply]
Yes, I used the `Recorder` class from the [rosbag2 repo](https://github.com/ros2/rosbag2). Not using the CLI but natively using the class in C++ spinning the node in an executor. I am doing it in a fork of this repo because of this [issue](https://github.com/ros2/rosbag2/issues/1188). The PR for the fork is [here](https://github.com/ros2/rosbag2/pull/1218#issuecomment-1369264889)

**Reply by james-rms** [reply]
The core issue here seems to be that the storage plugin is recording the wrong `.msg` definition for this message type. Could this be a problem with the new definition discovery logic?

**Reply by james-rms** [reply]
https://github.com/ros2/rosbag2/pull/1218/files#diff-278dec19a5b578bc59fb47ef35699a7d8ece72561c3cff9ba51ef78ae0193998R178 Most likely this is discovering `CompressedImage.msg` which ends with `Image.msg`.

**Reply by james-rms** [reply]
Looks like you need to include this patch: https://github.com/foxglove/ros-foxglove-bridge/pull/100/files

**Reply by aditya2592** [reply]
Thanks for confirming that and referring the fix! Let me include this patch and try again

**Reply by aditya2592** [reply]
It works after including the patch, so this issue can be closed
